src/Servers/Kestrel/Core/src/Internal/CancellationTokenSourcePool.cs

```diff
@@ -0,0 +1,65 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.Collections.Concurrent;
+using System.Threading;
+
+namespace Microsoft.AspNetCore.Server.Kestrel.Core.Internal
+{
+    internal class CancellationTokenSourcePool
+    {
+        private const int MaxQueueSize = 1024;
+
+        private readonly ConcurrentQueue<PooledCancellationTokenSource> _queue = new();
+        private int _count;
+
+        public PooledCancellationTokenSource Rent()
+        {
+            if (_queue.TryDequeue(out var cts))
+            {
+                Interlocked.Decrement(ref _count);
+                return cts;
+            }
+            return new PooledCancellationTokenSource(this);
+        }
+
+        private bool Return(PooledCancellationTokenSource cts)
+        {
+            // This counting isn't accurate, but it's good enough for what we need to avoid using _queue.Count which could be expensive
+            if (!cts.TryReset() || Interlocked.Increment(ref _count) > MaxQueueSize)
```

Comment:
We shouldn't call `Interlocked.Decrement(ref _count)` when `!cts.TryReset()`. That risks making the count negative.

```suggestion
            if (Interlocked.Increment(ref _count) > MaxQueueSize || !cts.TryReset())
```

Reply:
I don't love the "This counting isn't accurate," comment either. It's accurate enough to ensure `_queue` will never have more than `MaxQueueSize` items at any given time.

Reply:
Addressed the feedback, nice catch on the count.